src/box/iproto.cc

```diff
@@ -1800,12 +1965,30 @@ tx_process_call_on_yield(struct trigger *trigger, void *event)
 	return 0;
 }
 
+/**
+ * Because user can use iproto call request or iproto sql request to begin,
+ * commit, or rollback transaction, we must update all transaction related
+ * members of the stream structure after processing this type of requests.
+ */
+static void
+tx_set_txn_after(struct iproto_stream *stream, struct txn *prev_txn)
```

Comment:
Won't this do the same?
```Diff
@@ -1970,25 +1970,19 @@ tx_process_call_on_yield(struct trigger *trigger, void *event)
  * commit, or rollback transaction, we must update all transaction related
  * members of the stream structure after processing this type of requests.
  */
-static void
-tx_set_txn_after(struct iproto_stream *stream, struct txn *prev_txn)
+static inline void
+tx_set_txn_after(struct iproto_stream *stream)
 {
-	if (stream != NULL) {
-		stream->txn = in_txn();
-		if (prev_txn == NULL && stream->txn != NULL)
-			txn_detach();
-	}
+	if (stream != NULL && in_txn() != NULL)
+		stream->txn = txn_detach();
 }
 
 static void
 tx_process_call(struct cmsg *m)
 {
 	struct iproto_msg *msg = tx_accept_msg(m);
-	struct txn *prev_txn = NULL;
 	if (tx_check_schema(msg->header.schema_version))
 		goto error;
-	if (msg->stream != NULL)
-		prev_txn = msg->stream->txn;
 
 	/*
 	 * CALL/EVAL should copy its arguments so we can discard
@@ -2054,13 +2048,13 @@ tx_process_call(struct cmsg *m)
 		goto error;
 	}
 
-	tx_set_txn_after(msg->stream, prev_txn);
+	tx_set_txn_after(msg->stream);
 	iproto_reply_select(out, &svp, msg->header.sync,
 			    ::schema_version, count);
 	iproto_wpos_create(&msg->wpos, out);
 	return;
 error:
-	tx_set_txn_after(msg->stream, prev_txn);
+	tx_set_txn_after(msg->stream);
 	tx_reply_error(msg);
 }
 
@@ -2121,12 +2115,9 @@ tx_process_sql(struct cmsg *m)
 	const char *sql;
 	uint32_t len;
 	bool is_unprepare = false;
-	struct txn *prev_txn = NULL;
 
 	if (tx_check_schema(msg->header.schema_version))
 		goto error;
-	if (msg->stream != NULL)
-		prev_txn = msg->stream->txn;
 	assert(msg->header.type == IPROTO_EXECUTE ||
 	       msg->header.type == IPROTO_PREPARE);
 	tx_inject_delay();
@@ -2202,12 +2193,12 @@ tx_process_sql(struct cmsg *m)
 		goto error;
 	}
 	port_destroy(&port);
-	tx_set_txn_after(msg->stream, prev_txn);
+	tx_set_txn_after(msg->stream);
 	iproto_reply_sql(out, &header_svp, msg->header.sync, schema_version);
 	iproto_wpos_create(&msg->wpos, out);
 	return;
 error:
-	tx_set_txn_after(msg->stream, prev_txn);
+	tx_set_txn_after(msg->stream);
 	tx_reply_error(msg);
 }
```

Reply:
No, we can commit in `call` or `eval`, so we need to set stream->txn all time, also i think it's not good to detach transaction many time.

Reply:
It seems to me you need to make `txn_detach()` work when there is no a transaction and return `NULL` then. It could make this code simpler and maybe some other places too.

Reply:
Change `txn_detach`, now it's work with in_txn() == NULL, also ti's clears both triggers, and i call it from
tx_end_msg all time.